File: iOS/Sources/Presentation/ViewModels/PocketDetailViewModel.swift
import Foundation
import Combine

@MainActor
class PocketDetailViewModel: ObservableObject {
    @Published var pocket: Pocket
    @Published var budgets: [Budget] = []
    @Published var errorMessage: String?
    @Published var isLoading = false

    private let pocketInteractor: PocketInteractor
    private let budgetInteractor: BudgetInteractor

    init(pocket: Pocket, pocketInteractor: PocketInteractor, budgetInteractor: BudgetInteractor) {
        self.pocket = pocket
        self.pocketInteractor = pocketInteractor
        self.budgetInteractor = budgetInteractor
    }

    var totalAllocated: Double {
        budgets.reduce(0) { $0 + $1.allocatedAmount }
    }

    var totalSpent: Double {
        budgets.reduce(0) { $0 + $1.spentAmount }
    }

    func loadBudgets() async {
        isLoading = true
        do {
            budgets = try await budgetInteractor.getBudgets(forPocketId: pocket.id)
        } catch {
            errorMessage = error.localizedDescription
        }
        isLoading = false
    }

    func refreshPocket() async {
        do {
            pocket = try await pocketInteractor.getPocket(id: pocket.id)
        } catch {
            errorMessage = error.localizedDescription
        }
    }

    func addFunds(amount: Double) async -> Bool {
        do {
            pocket = try await pocketInteractor.addFunds(id: pocket.id, amount: amount)
            return true
        } catch {
            errorMessage = error.localizedDescription
            return false
        }
    }

    func deleteBudget(at offsets: IndexSet) async {
        for index in offsets {
            let budget = budgets[index]
            do {
                try await budgetInteractor.deleteBudget(id: budget.id)
                budgets.remove(at: index)
                await refreshPocket()
            } catch {
                errorMessage = error.localizedDescription
            }
        }
    }
}

File: android/app/src/main/java/com/budgetmanager/ui/viewmodel/PocketDetailViewModel.kt
package com.budgetmanager.ui.viewmodel

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.budgetmanager.domain.entity.Budget
import com.budgetmanager.domain.entity.Pocket
import com.budgetmanager.domain.interactor.BudgetInteractor
import com.budgetmanager.domain.interactor.PocketInteractor
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch

data class PocketDetailState(
    val pocket: Pocket? = null,
    val budgets: List<Budget> = emptyList(),
    val isLoading: Boolean = false,
    val errorMessage: String? = null
) {
    val totalAllocated: Double
        get() = budgets.sumOf { it.allocatedAmount }

    val totalSpent: Double
        get() = budgets.sumOf { it.spentAmount }
}

class PocketDetailViewModel(
    private val pocketId: Long,
    private val pocketInteractor: PocketInteractor,
    private val budgetInteractor: BudgetInteractor
) : ViewModel() {

    private val _state = MutableStateFlow(PocketDetailState())
    val state: StateFlow<PocketDetailState> = _state.asStateFlow()

    fun loadData() {
        viewModelScope.launch {
            _state.value = _state.value.copy(isLoading = true, errorMessage = null)
            try {
                val pocket = pocketInteractor.getPocket(pocketId)
                val budgets = budgetInteractor.getBudgetsByPocketId(pocketId)
                _state.value = _state.value.copy(
                    pocket = pocket,
                    budgets = budgets,
                    isLoading = false
                )
            } catch (e: Exception) {
                _state.value = _state.value.copy(
                    isLoading = false,
                    errorMessage = e.message ?: "Failed to load data"
                )
            }
        }
    }

    fun addFunds(amount: Double) {
        viewModelScope.launch {
            try {
                pocketInteractor.addFunds(pocketId, amount)
                loadData()
            } catch (e: Exception) {
                _state.value = _state.value.copy(errorMessage = e.message)
            }
        }
    }

    fun deleteBudget(id: Long) {
        viewModelScope.launch {
            try {
                budgetInteractor.deleteBudget(id)
                loadData()
            } catch (e: Exception) {
                _state.value = _state.value.copy(errorMessage = e.message)
            }
        }
    }

    fun clearError() {
        _state.value = _state.value.copy(errorMessage = null)
    }
}
